File: python/day-3a.py
def get_priority(char):
    val = ord(char)
    if val > 96:
        return val - 96
    return val - 65 + 27


def main():
    with open('input/day-3.txt', 'r') as f:
        total = 0

        for line in f.readlines():
            if line.endswith('\n'):
                line = line[:-1]
            letters = {}
            mid_point = int(len(line) / 2)

            for c in line[:mid_point]:
                letters[c] = True

            for c in line[mid_point:]:
                if c in letters:
                    # print(c, get_priority(c))
                    total += get_priority(c)
                    break

        return total


print(main())

File: rust/src/bin/day-3a.rs
use std::fs::read_to_string;
use std::collections::HashMap;

fn get_priority(c: char) -> u32 {
    let val = c as u32;
    if val > 'a' as u32 {
        return val - 'a' as u32 + 1;
    }
    return val - 'A' as u32 + 27;
}

fn get_line_val(line: &str) -> u32 {
    let mut letters: HashMap<char, bool> = HashMap::with_capacity(52);
    let mid_point: usize = line.len() >> 1;
    let mut chars = line.chars();
    
    for _ in 0..mid_point {
        if let Some(c) = chars.next() {
            letters.insert(c, true);
        }
    }

    for _ in mid_point..line.len() {
        if let Some(c) = chars.next() {
            if let Some(_) = letters.get(&c) {
                // println!("{}", get_priority(c));
                return get_priority(c);
            }
        }
    }

    return 0;
}

fn main() -> std::io::Result<()> {
    let total: u32 = read_to_string("day-3a/input.txt")
        .unwrap()
        .lines()
        .map(get_line_val)
        .sum();

    println!("{}", total);
    Ok(())
}
